File: ios/CruiseMesh/Mesh/LanEndpointSender.swift
import Foundation
import os.log

private let lanEndpointHintLifetimeMs: Int64 = 15 * 60 * 1_000

enum LanEndpointSender {
    private static let log = Logger(subsystem: "com.cruisemesh", category: "LanEndpointSender")

    static func queueToAllCapableContacts(
        store: MessageStore,
        identity: Identity,
        endpoint: LanManualEndpoint,
        instanceToken: Data,
        networkId: String?
    ) {
        guard let networkId else { return }
        for contact in (try? store.listContacts()) ?? [] where LanCapabilityStore.isSupported(userId: contact.userId) {
            queueToContact(
                store: store,
                identity: identity,
                contact: contact,
                endpoint: endpoint,
                instanceToken: instanceToken,
                networkId: networkId
            )
        }
    }

    static func queueToContact(
        store: MessageStore,
        identity: Identity,
        contact: Contact,
        endpoint: LanManualEndpoint,
        instanceToken: Data,
        networkId: String
    ) {
        guard LanCapabilityStore.shouldSendEndpoint(
            userId: contact.userId,
            networkId: networkId,
            endpoint: endpoint,
            instanceToken: instanceToken
        ) else { return }

        let timestamp = Int64(Date().timeIntervalSince1970 * 1_000)
        let payload: Data
        do {
            payload = try encodeLanEndpointContent(content: LanEndpointContent(
                instanceToken: instanceToken,
                networkId: Data(networkId.utf8),
                host: endpoint.host,
                port: endpoint.port,
                expiresAtMs: timestamp + lanEndpointHintLifetimeMs
            ))
        } catch {
            log.warning("Unable to encode sealed LAN endpoint hint")
            return
        }
        guard let authored = try? store.authorPairwiseMessage(
            identity: identity,
            contact: contact,
            kind: ProtocolKind.lanEndpointHint,
            payload: payload,
            replyToMsgId: nil,
            timestampMs: timestamp
        ) else { return }
        GossipState.seenIds.record(msgId: authored.envelope.msgId)
        RelaySyncEvents.requestSync()
        _ = MeshRouter.sendToUserId(userId: contact.userId, frame: authored.frame)
    }
}

File: android/app/src/main/kotlin/com/cruisemesh/app/mesh/LanEndpointSender.kt
package com.cruisemesh.app.mesh

import android.content.Context
import android.util.Log
import com.cruisemesh.app.chat.UserIdHex
import uniffi.cruisemesh_core.Contact
import uniffi.cruisemesh_core.Frame
import uniffi.cruisemesh_core.Identity
import uniffi.cruisemesh_core.LanEndpointContent
import uniffi.cruisemesh_core.MessageStore
import uniffi.cruisemesh_core.encodeLanEndpointContent

private const val TAG = "LanEndpointSender"
// KIND_LAN_ENDPOINT_HINT comes from MeshWireConstants.kt (FA15).
private const val HINT_LIFETIME_MS = 15 * 60 * 1_000L

internal object LanEndpointSender {
    fun queueToAllCapableContacts(
        context: Context,
        store: MessageStore,
        identity: Identity,
        hint: Frame.LanEndpoint,
        networkId: String?,
    ) {
        if (networkId == null) return
        for (contact in store.listContacts()) {
            if (!LanCapabilityStore.isSupported(context, contact.userId)) continue
            queueToContact(context, store, identity, contact, hint, networkId)
        }
    }

    fun queueToContact(
        context: Context,
        store: MessageStore,
        identity: Identity,
        contact: Contact,
        hint: Frame.LanEndpoint,
        networkId: String,
    ) {
        if (
            !LanCapabilityStore.shouldSendEndpoint(
                context,
                contact.userId,
                networkId,
                hint.host,
                hint.port.toInt(),
                hint.instanceToken,
            )
        ) {
            return
        }
        val timestamp = System.currentTimeMillis()
        val payload = try {
            encodeLanEndpointContent(
                LanEndpointContent(
                    instanceToken = hint.instanceToken,
                    networkId = networkId.toByteArray(Charsets.UTF_8),
                    host = hint.host,
                    port = hint.port,
                    expiresAtMs = timestamp + HINT_LIFETIME_MS,
                ),
            )
        } catch (error: Exception) {
            Log.w(TAG, "Unable to encode sealed LAN endpoint hint", error)
            return
        }
        val authored = store.authorPairwiseMessage(
            identity,
            contact,
            KIND_LAN_ENDPOINT_HINT,
            payload,
            null,
            timestamp,
        )
        GossipState.seenIds.record(authored.envelope.msgId)
        RelaySyncEvents.requestSync()
        if (!MeshRouter.sendToUserId(contact.userId, authored.frame)) {
            Log.i(
                TAG,
                "Queued sealed LAN endpoint for ${UserIdHex.encode(contact.userId)}",
            )
        }
    }
}
